Added scripts for the 5-2 petition followup mailing.

aoby opened the pull request.

ashercodeorg commented on bin/mail/2016-05-02-petition-followup/generate-list.rb:
```diff
@@ -0,0 +1,45 @@
+# First, get all US teachers and past organizers.
+teacher_query = %q(
+  ((kind_s:user && role_s:teacher) ||
+  kind_s:HocSignup2015 || kind_s:HocSignup2014 || kind_s:"CSEdWeekEvent2013") && (
```
nit: Is there a reason for the quotes around CSEdWeekEvent2013 but not the others?

aoby replied:
copy/paste. I'll remove the quotes.

ashercodeorg commented on bin/mail/2016-05-02-petition-followup/generate-list.rb:
```diff
@@ -0,0 +1,45 @@
+
+
+# Next, filter for those who did not click through the previous email.
+PREVIOUS_MESSAGE_NAME = '4-22-petition-congress-3b-no'
```
Does this mean we will automatically be resending to the 4.8% (the experimentals) that did click through?

aoby replied:
No, because this only applies to the final email template '4-22-petition-congress-3b-no'. But you raise a good point. We will not be re-sending to any of the experimental emails. I'll follow up with marketing to see if we should do those too.

ashercodeorg commented on bin/mail/2016-05-02-petition-followup/generate-list.rb:
```diff
@@ -0,0 +1,45 @@
+message_id = DB[:poste_messages].where(name: PREVIOUS_MESSAGE_NAME).first[:id]
+url_id = DB[:poste_urls].where(url: CLICK_THROUGH_URL).first[:id]
+
+# sending the entire email list was causing problems, so splitting into smaller lists
```
sending or querying?

aoby replied:
including in the query, as `contact_email IN (...)` I'll update this comment.

ashercodeorg commented on bin/mail/2016-05-02-petition-followup/send-test-mails:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+../send-to-mailing-list 5-2-petition-follow-up ../mailing-common/test-params-us.csv
```
Q: Is the purpose of this to test the template and make sure the email is populated correctly from the information in the CSV?

aoby replied:
Yes, and that it's formatted correctly. I always like to send myself all variations of the emails first to check formatting before sending them out to a mass audience.

ashercodeorg commented on bin/mail/2016-05-02-petition-followup/generate-list.rb:
```diff
@@ -0,0 +1,44 @@
+    WHERE poste_deliveries.message_id = #{message_id} AND
+      poste_clicks.id IS NULL AND
+      contact_email IN (#{email_list})
+  SQL
```
Is it sensical to, as a test, run the same query with an INNER JOIN to confirm that some records are being queried away? Note that a typo in the CLICK_THROUGH_URL, for example, would not be super apparent.

aoby replied:
Quite. That's a good suggestion. I'll add it.

aoby commented:
@ashercodeorg I updated the PR with your suggestions. After consulting with Marketing, i expanded the query to include the experiment emails too. We are sticking with everyone who didn't click through regardless of whether they opened the original email. 

I added a test query and manually verified I'd get an error for both conditions: completely faulty url, or a valid but incorrect url (i.e. one used in a previous email campaign).

ashercodeorg commented on bin/mail/2016-05-02-petition-followup/generate-list.rb:
```diff
@@ -0,0 +1,61 @@
+# Next, filter for those who did not click through any of the previous petition emails (experiments or final).
+PREVIOUS_MESSAGE_NAME_PREFIX = '4-22-petition-congress%'
+CLICK_THROUGH_URL = 'http://bit.ly/computersciencepetition'
+message_id_list = DB[:poste_messages].where(Sequel.like(:name, PREVIOUS_MESSAGE_NAME_PREFIX)).map(:id).join(',')
```
nit: Long line. Perhaps

```
message_id_list = DB[:poste_messages].
  where(Sequel.like(:name, PREVIOUS_MESSAGE_NAME_PREFIX)).
  map(:id).
  join(',')
```

ashercodeorg commented on bin/mail/2016-05-02-petition-followup/generate-list.rb:
```diff
@@ -0,0 +1,61 @@
+  unless test_query_succeeded
+    test_query = <<-SQL
+      SELECT COUNT(0) AS click_count FROM poste_deliveries
+      INNER JOIN poste_clicks ON (poste_clicks.delivery_id = poste_deliveries.id AND poste_clicks.url_id = #{url_id})
```
nit: Long line. Perhaps

```
      INNER JOIN poste_clicks ON (poste_clicks.delivery_id = poste_deliveries.id
          AND poste_clicks.url_id = #{url_id})
```

ashercodeorg commented on bin/mail/2016-05-02-petition-followup/generate-list.rb:
```diff
@@ -0,0 +1,61 @@
+
+  filter_query = <<-SQL
+    SELECT contact_email FROM poste_deliveries
+    LEFT JOIN poste_clicks ON (poste_clicks.delivery_id = poste_deliveries.id AND poste_clicks.url_id = #{url_id})
```
nit: Long line.

FWIW, I believe the parentheses are unnecessary.

ashercodeorg commented on bin/mail/2016-05-02-petition-followup/generate-list.rb:
```diff
@@ -0,0 +1,61 @@
+      WHERE poste_deliveries.message_id IN (#{message_id_list}) AND
+        contact_email IN (#{email_list})
+    SQL
+    test_query_succeeded = true if DB.fetch(test_query).get(:click_count) > 0
```
This could be simplified to
`test_query_succeeded ||= DB.fetch(test_query).get(:click_count) > 0`
or even
`test_query_succeeded = DB.fetch(test_query).get(:click_count) > 0`
since it is only run if `!test_query_succeeded`.

ashercodeorg commented:
Bunch of nits, overall LGTM. Note that the ci/circleci failure seems to be spurious (a JS error).

aoby commented:
I updated with your suggestions. The unrelated js error has been fixed in https://github.com/code-dot-org/code-dot-org/pull/8163, so I rebased and am running the tests again.
